### PR TITLE
Add Support for Partition Backup and Restores

### PR DESCRIPTION
**Pull request by riptidewave93**
This adds the ability to do system backups and restores using the flag files full_backup and full_restore, as well as the functions backup_mtd_partition and restore_mtd_partition which can be called from a image file.

Before this is merged, https://github.com/tchebb/eureka_buildroot/pull/2 must be merged upstream otherwise these features will not function properly.

### CODE REVIEW

**Comment by tchebb** [comment]
Please rebase on top of the latest changes. GitHub doesn't seem to let me add line comments until you do.

**Comment by riptidewave93** [comment]
Rebase Done

File: bundled-mods/full-restore/imager.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+set -e
+
+if test -z "$1" ; then
+        fatal "No Backup Folder Defined in full_backup"
```
Comment (tchebb) [inline bundled-mods/full-restore/imager.sh]: `full_restore`, not `full_backup`, right?

File: bundled-mods/full-backup/imager.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+set -e
+
+backup_mtd_partition 'rootfs'
+
```
Comment (tchebb) [inline bundled-mods/full-backup/imager.sh]: No need for this whitespace between the `backup_mtd_partition` calls. It just makes the file harder to read.

File: bundled-mods/full-restore/imager.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+fi
+
+# $1 should be the folder number of your backup, ex 1 for ./backup-1
+# it should NOT be a full path!
+
+restore_mtd_partition 'rootfs' "$1"
+
```
Comment (tchebb) [inline bundled-mods/full-restore/imager.sh]: See full-backup/imager.sh comments.

File: src/flash-image [hunk 4]
```diff
@@ -81,6 +81,157 @@ mtd_lookup() {
...
+# $1: The name (NOT number) of the MTD partition to dump.
+# Returns: 0 on success, 1 on failure.
+backup_mtd_partition() {
+	_PARTITION="$(mtd_lookup "$1")"
+	if test "$?" -ne 0 ; then
+		return 1;
+	fi
+	
+	# I know this is bad, but we can't store this in /tmp if images are there due to space
+	_TEMPMOUNT="$(mktemp -d)"
```
Comment (tchebb) [inline src/flash-image]: Use `_TEMPMOUNT="${RUNTIME_DIR}/backups"` and remove all the mounting (and cleanup) logic.

Comment (tchebb) [inline src/flash-image]: Also, `_TEMP_MOUNT` would be a preferable name for consistency.

Comment (tchebb) [inline src/flash-image]: Ignore my first comment on this line for now; I forgot to export `$RUNTIME_DIR` so it'll fail. Will fix later today.

File: src/flash-image [hunk 5]
```diff
@@ -81,6 +81,157 @@ mtd_lookup() {
...
+			fi
+		fi
+	fi
+
+	#Do our own Cleanup
+	umount ${_TEMPMOUNT} && rm -r ${_TEMPMOUNT}
+}
+
+# restore_mtd_partition
+# Used to Restore dumps bade by backup_mtd_partition.
```
Comment (tchebb) [inline src/flash-image]: s/bade/made/

File: src/flasher [hunk 6]
```diff
@@ -46,6 +46,15 @@ select_sequential_dir() {
 	echo "${1}/${I}"
 }
 
+unmount_images() {
```
Comment (tchebb) [inline src/flasher]: Why does this function get moved? Even if it should be moved, that should go in a separate pull request as it has no bearing on this functionality.

File: src/flash-image [hunk 7]
```diff
@@ -81,6 +81,157 @@ mtd_lookup() {
...
+	fi
+	_BACKUPFOLDER="$(cat /tmp/backup_dir)"
+        
+        # are we data? otherwise, use nanddump.
+        #
+        # We do this because dd & nanddump...
+        # 1: uses ALOT less space
+        # 2: dumps of this partition do not write back properly
+	if test "${1}" = "userdata" ; then
+		dry_run "Would backup ${1} to $_TEMPMOUNT/$_BACKUPFOLDER/${1}.zip"
```
Comment (tchebb) [inline src/flash-image]: Nice job supporting `$DRY_RUN`.

**Comment by riptidewave93** [comment]
Fixes added, the moving of the function in flasher was to keep functions at the top of the file. I will submit another pull request with that fix once this gets merged.

File: src/flash-image [hunk 8]
```diff
@@ -81,6 +81,157 @@ mtd_lookup() {
...
+	# I know this is bad, but we can't store this in /tmp if images are there due to space
+	_TEMP_MOUNT="$(mktemp -d)"
+	if ! mount "/dev/sda1" "$_TEMP_MOUNT" ; then
+		log "Could not mount /dev/sda1 to $_TEMP_MOUNT"
+		rm -r ${_TEMP_MOUNT}
+		return 1;
+	fi
+
+	# Do we know the mount point? We do this so all backups go to the same folder
+	if ! test -f "/tmp/backup_dir" ; then
```
Comment (tchebb) [inline src/flash-image:105]: Kill all the `$_TEMP_MOUNT` and `/tmp/backup_dir` stuff. `flasher` now handles all that for you. Just drop backups into `${RUNTIME_DIR}/backups` and you'll be all set.

**Comment by tchebb** [comment]
Pulled into local branch `backup`, where I'll make the final changes and merge to master. Thanks for all the work on this.
